src/github/webhook.rs

```diff
@@ -0,0 +1,337 @@
+    Push,
+    /// A Git branch or tag is created.
+    ///
+    /// This gets translated to [`github::Event::Create`] when sent to a handler.
+    ///
+    /// <https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#create>
+    Create,
+    /// All other unhandled webhooks.
+    Other,
+}
+
+impl std::str::FromStr for EventName {
+    type Err = std::convert::Infallible;
+    fn from_str(s: &str) -> Result<EventName, Self::Err> {
+        Ok(match s {
+            "pull_request_review" => EventName::PullRequestReview,
+            "pull_request_review_comment" => EventName::PullRequestReviewComment,
+            "issue_comment" => EventName::IssueComment,
+            "pull_request" => EventName::PullRequest,
+            "issues" => EventName::Issue,
+            "push" => EventName::Push,
+            "create" => EventName::Create,
+            _ => EventName::Other,
+        })
+    }
+}
+
+impl fmt::Display for EventName {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                EventName::PullRequestReview => "pull_request_review",
+                EventName::PullRequestReviewComment => "pull_request_review_comment",
+                EventName::IssueComment => "issue_comment",
+                EventName::Issue => "issues",
+                EventName::PullRequest => "pull_request",
+                EventName::Push => "push",
+                EventName::Create => "create",
+                EventName::Other => "other",
+            }
+        )
+    }
+}
+
+#[derive(Debug)]
+pub struct WebhookError(
+    #[allow(dead_code)] // Used in debug display
+    anyhow::Error,
+);
+
+impl From<anyhow::Error> for WebhookError {
+    fn from(e: anyhow::Error) -> WebhookError {
+        WebhookError(e)
+    }
+}
+
+pub fn deserialize_payload<T: serde::de::DeserializeOwned>(v: &str) -> anyhow::Result<T> {
+    let mut deserializer = serde_json::Deserializer::from_str(&v);
+    let res: Result<T, _> = serde_path_to_error::deserialize(&mut deserializer);
+    match res {
+        Ok(r) => Ok(r),
+        Err(e) => {
+            tracing::error!("failed to deserialize webhook payload: {v}");
+            let ctx = format!("at {:?}", e.path());
+            Err(e.into_inner()).context(ctx)
+        }
+    }
+}
+
+pub async fn webhook(
+    headers: HeaderMap,
+    State(ctx): State<Arc<crate::handlers::Context>>,
+    body: Bytes,
+) -> axum::response::Response {
+    // Extract X-GitHub-Event header
+    let Some(ev) = headers.get("X-GitHub-Event") else {
+        return (StatusCode::BAD_REQUEST, "X-GitHub-Event header must be set").into_response();
+    };
+    let Ok(ev) = ev.to_str() else {
+        return (
+            StatusCode::BAD_REQUEST,
+            "X-GitHub-Event header must be UTF-8 encoded",
+        )
+            .into_response();
+    };
+    let Ok(event) = ev.parse::<EventName>();
```

Comment:
I'm surprised this syntax works. O.o

Reply:
That's because there isn't really a error type, it's a irrefutable error.

https://github.com/Urgau/triagebot/blob/28af06f1926b14a2c931e3052d2c8ed71f4571a9/src/github/webhook.rs#L63